### PR TITLE
[HotA] Bank support

### PR DESCRIPTION
**Pull request by Laserlicht**
Bank support for HotA 1.7

add in config: `"type" : "bank",` to `special1`

### CODE REVIEW

File: CCallback.cpp [hunk 1]
```diff
@@ -217,6 +217,16 @@ bool CCallback::buildBuilding(const CGTownInstance *town, BuildingID buildingID)
 	return true;
 }
 
+bool CCallback::triggerTownSpecialBuildingAction(const CGTownInstance *town, BuildingSubID::EBuildingSubID subBuildingID)
+{
+	if(town->tempOwner!=player)
+		return false;
+
```
Comment (IvanSavenko) [inline CCallback.cpp:224]: Should probably be handled in UI?

File: CCallback.cpp [hunk 2]
```diff
@@ -217,6 +217,16 @@ bool CCallback::buildBuilding(const CGTownInstance *town, BuildingID buildingID)
 	return true;
 }
 
+bool CCallback::triggerTownSpecialBuildingAction(const CGTownInstance *town, BuildingSubID::EBuildingSubID subBuildingID)
```
Comment (IvanSavenko) [inline CCallback.cpp:220]: We might also want to extend this logic to mana whirlpool, similar to what hota did with it.
Or even allow it for any building, e.g. flag in building config, so you can make explicitly visitable stables if you want to

File: client/windows/CCastleInterface.cpp [hunk 3]
```diff
@@ -1054,6 +1058,21 @@ void CCastleBuildings::enterAnyThievesGuild()
 	LOCPLINT->showInfoDialog(CGI->generaltexth->translate("vcmi.adventureMap.noTownWithTavern"));
 }
 
+void CCastleBuildings::enterBank()
+{
+	std::vector<std::shared_ptr<CComponent>> components;
+	if(town->bonusValue.second > 0)
+	{
+		components.push_back(std::make_shared<CComponent>(ComponentType::RESOURCE, GameResID(GameResID::GOLD), "-500/" + CGI->generaltexth->translate("core.genrltxt.64")));
```
Comment (IvanSavenko) [inline client/windows/CCastleInterface.cpp:1066]: There is `ComponentType::RESOURCE_PER_DAY` for this purpose if I understand correctly

File: server/CGameHandler.cpp [hunk 4]
```diff
@@ -2469,6 +2469,30 @@ bool CGameHandler::buildStructure(ObjectInstanceID tid, BuildingID requestedID,
...
 }
 
+bool CGameHandler::triggerTownSpecialBuildingAction(ObjectInstanceID tid, BuildingSubID::EBuildingSubID sid)
+{
+	const CGTownInstance * t = getTown(tid);
+
+	bool hasBuilding = false;
+	for (auto building : t->town->buildings)
+		if(vstd::contains(t->builtBuildings, building.first) && building.second->subId == sid)
+			hasBuilding = true;
```
Comment (IvanSavenko) [inline server/CGameHandler.cpp:2479]: use `CTown::getBuildingType` instead?

File: CCallback.cpp (hunk 1 above)
Comment (Laserlicht) [reply CCallback.cpp:224]: adapted this logic from `CCallback::buildBuilding`

File: server/CGameHandler.cpp [hunk 5]
```diff
@@ -2469,6 +2469,30 @@ bool CGameHandler::buildStructure(ObjectInstanceID tid, BuildingID requestedID,
...
+		return false;
+
+	if(sid == BuildingSubID::EBuildingSubID::BANK)
+	{
+		TResources res;
+		res[EGameResID::GOLD] = 2500;
+		giveResources(t->getOwner(), res);
+
+		setObjPropertyValue(t->id, ObjProperty::BONUS_VALUE_SECOND, 2500);
+	}
```
[the remark this answers is not in the document]
Comment (Laserlicht) [reply server/CGameHandler.cpp:2491]: Maybe rather an task for later refactoring?
It already contains handling of several BuildingSubID's and BuildingID's.

File: CCallback.cpp (hunk 2 above)
Comment (Laserlicht) [reply CCallback.cpp:220]: Yes, there's nothing wrong with extending it?

File: server/CGameHandler.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (IvanSavenko) [reply server/CGameHandler.cpp:2514]: Will probably do this myself in separate PR

File: CCallback.cpp (hunk 2 above)
Comment (IvanSavenko) [reply CCallback.cpp:220]: Yeah, should be fine allowing this for all buildings.

Explicit visit can also be used to fix our old feature-bug where stables & such are visited on new week automatically, without even opening town screen

Comment (dydzio0614) [reply CCallback.cpp:220]: well there are 3 types of visit that make sense:
1. automatic (like now) - it is kinda QoL that buildings like stables work without opening town screen
2. on opening town screen (original h3 mana vortex)
3. hota 1.7 has its new manual handling of mana vortex

Comment (IvanSavenko) [reply CCallback.cpp:220]: > automatic (like now) - it is kinda QoL 

Yeah, but sometimes result in unwanted message spam on new week, so kind of mixed feelings. Perhaps alternative approach would be to hide these messages.

**Pull request merged**